pkg/etcdenvvar/envvarcontroller.go

```diff
@@ -128,15 +130,53 @@ func (c *EnvVarController) checkEnvVars() error {
 		return err
 	}
 
-	currEnvVarMap, err := getEtcdEnvVars(envVarContext{
+	ctx := envVarContext{
 		targetImagePullSpec:  c.targetImagePullSpec,
 		spec:                 *operatorSpec,
 		status:               *operatorStatus,
 		configmapLister:      c.configmapLister,
 		nodeLister:           c.nodeLister,
 		infrastructureLister: c.infrastructureLister,
 		networkLister:        c.networkLister,
-	})
+	}
+
+	bootstrapComplete, err := isBootstrapComplete(c.configmapLister, c.operatorClient)
+	if err != nil {
+		return fmt.Errorf("couldn't determine bootstrap status: %w", err)
+	}
+	isUnsupportedUnsafeEtcd, err := ceohelpers.IsUnsupportedUnsafeEtcd(&ctx.spec)
+	if err != nil {
+		return fmt.Errorf("couldn't determine etcd unsupported override status: %w", err)
+	}
+	// TODO: determine this through a state check
+	isManagedByAssistedInstaller := true
+	nodeCount := len(ctx.status.NodeStatuses)
+
+	// There are three discrete validation paths to enforce HA invariants
+	// depending on whether the unsupported/unsafe config is set, whether the
+	// cluster is managed by assisted installer, and the default configuration.
+	switch {
+	case isUnsupportedUnsafeEtcd:
+		// When running in an unsupported/unsafe configuration, no guarantees are
+		// made once a single node is available.
+		if nodeCount < 1 {
+			return fmt.Errorf("at least one node is required to have a valid configuration")
+		}
+	case isManagedByAssistedInstaller:
```

Comment:
if somehow both `isUnsupportedUnsafeEtcd` and `isManagedByAssistedInstaller` are passed I think this should be a net fail long term. Short term we tolerate? Should the order be changed so we are testing `isManagedByAssistedInstaller` always? My understanding is they will send both during the transition.

Reply:
By net fail I mean we literally go degraded is we observe both as our 4.7 shipped solution?

Reply:
Hm. Good questions. Maybe it depends on whether we consider these orthogonal? The unsupported has clear semantics on its own and I can see an argument turning on the unsupported flag should take precedence.

I'm not sure I understand re: transition — shouldn't unsupported override be removed before GA? Consider this order of operations:

1. Test this PR against a branch of assisted installer which removes the unsupported override
2. Merge this PR (as it should have no impact/regressions on existing platforms)
3. Merge changes to assisted installer to remove the unsupported override

If the unsupported override remains enabled, I would expect the cluster to be marked tainted somehow no matter what logic we have in this particular controller as a matter of generic platform behavior. Is that not the case? If there's no generic cluster tainting logic that will pick up on the presence of the unsupported config override, then I could understand covering the case here. Any insights onto how that works? cc @deads2k

Reply:
For now, I went ahead and added a check which should cause us to fail if assisted installer is running with the unsupported override. This should help prove things work in the supported configuration. We can remove the check after testing if we want.